substrate/client/consensus/beefy/src/communication/fisherman.rs

```diff
@@ -0,0 +1,244 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use crate::{
+	error::Error,
+	justification::BeefyVersionedFinalityProof,
+	keystore::{BeefyKeystore, BeefySignatureHasher},
+	LOG_TARGET,
+};
+use log::debug;
+use sc_client_api::Backend;
+use sp_api::ProvideRuntimeApi;
+use sp_blockchain::HeaderBackend;
+use sp_consensus_beefy::{
+	check_fork_equivocation_proof,
+	ecdsa_crypto::{AuthorityId, Signature},
+	BeefyApi, ForkEquivocationProof, Payload, PayloadProvider, SignedCommitment, ValidatorSet,
+	VoteMessage,
+};
+use sp_runtime::{
+	generic::BlockId,
+	traits::{Block, Header, NumberFor},
+};
+use std::{marker::PhantomData, sync::Arc};
+
+pub(crate) trait BeefyFisherman<B: Block>: Send + Sync {
```

Comment:
nice idea - done in d18897d05c8

Note this muddies the gossip unit tests in 30ea46ced80 a bit since can't just phantom dummy anymore and the unit tests also require instances of `Backend`, `TestApi`, etc., but overall it's still simpler than keeping the `BeefyFisherman` trait around.
One could move the affected unit tests ([should_validate_messages(...)](https://github.com/lederstrumpf/polkadot-sdk/blob/0283226cca69d659b0a217e34fc7ef17cb4cc860/substrate/client/consensus/beefy/src/communication/gossip.rs#L586), [messages_allowed_and_expired(...)](https://github.com/lederstrumpf/polkadot-sdk/tree/0283226cca69d659b0a217e34fc7ef17cb4cc860/substrate/client/consensus/beefy/src/communication/gossip.rs#L707), [messages_rebroadcast(...)](https://github.com/lederstrumpf/polkadot-sdk/tree/0283226cca69d659b0a217e34fc7ef17cb4cc860/substrate/client/consensus/beefy/src/communication/gossip.rs#L791)) to the integration tests since they now depend on the same integrations, but since they don't actually test the integrations (they only require them given type constraints), on balance I think they're more appropriate to remain where they are.